### PR TITLE
Future compatibility with multi-hop routing

### PR DESCRIPTION
**Pull request by cwgoes**
Ref https://github.com/cosmos/ics/issues/138#issuecomment-509029441

### CODE REVIEW

**Comment by cwgoes** [comment]
- [ ] Loopback
- [ ] Avoid duplicate key / value data on channels
- [ ] Read through logic again

**Comment by cwgoes** [comment]
Split off loopback - https://github.com/cosmos/ics/issues/151

File: spec/ics-004-channel-and-packet-semantics/README.md [hunk 1]
```diff
@@ -423,20 +430,21 @@ Note that the full packet is not stored in the state of the chain - merely a sho
...
   assert(channel.state === OPEN)
-  assert(authenticate(get(portKey(channel.portIdentifier))))
+  assert(authenticate(get(portKey(packet.sourcePort))))
+  assert(packet.destPort === channel.counterpartyPortIdentifier)
   assert(packet.destChannel === channel.counterpartyChannelIdentifier)
-  connection = get(connectionKey(packet.sourceConnection))
+  connection = get(connectionKey(packet.connectionHops[0]))
   assert(connection.state === OPEN)
-  assert(packet.destConnection === connection.counterpartyConnectionIdentifier)
+  assert(packet.connectionHops === channel.connectionHops)
```
Comment (ancazamfir) [inline spec/ics-004-channel-and-packet-semantics/README.md]: Will this be true with number of connection hops > 2? Is the plan to have all hops in the packet?

File: spec/ics-004-channel-and-packet-semantics/README.md [hunk 2]
```diff
@@ -67,18 +67,19 @@ interface ChannelEnd {
...
   counterpartyChannelIdentifier: Identifier
-  portIdentifier: Identifier
   counterpartyPortIdentifier: Identifier
+  connectionHops: [Identifier]
   nextTimeoutHeight: uint64
 }
 ```
 
 - The `state` is the current state of the channel end.
+- The `ordering` field indicates whether the channel is ordered or unordered.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply spec/ics-004-channel-and-packet-semantics/README.md]: Updated.

File: spec/ics-004-channel-and-packet-semantics/README.md [hunk 3]
```diff
@@ -67,18 +67,19 @@ interface ChannelEnd {
   state: ChannelEndState
   ordering: ChannelOrder
   counterpartyChannelIdentifier: Identifier
-  portIdentifier: Identifier
   counterpartyPortIdentifier: Identifier
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply spec/ics-004-channel-and-packet-semantics/README.md]: OK; updated.

File: spec/ics-004-channel-and-packet-semantics/README.md [hunk 4]
```diff
@@ -158,28 +160,28 @@ The architecture of clients, connections, channels and packets:
 Channel structures are stored under a store key prefix unique to a combination of a connection identifier and channel identifier:
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply spec/ics-004-channel-and-packet-semantics/README.md]: Thanks.

File: spec/ics-004-channel-and-packet-semantics/README.md [hunk 5]
```diff
@@ -212,47 +214,52 @@ could be implemented to provide this).
 
 ```typescript
 function chanOpenInit(
-  order: ChannelOrder, connectionIdentifier: Identifier, channelIdentifier: Identifier,
+  order: ChannelOrder, connectionHops: [Identifier], channelIdentifier: Identifier,
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply spec/ics-004-channel-and-packet-semantics/README.md]: Updated.

File: spec/ics-004-channel-and-packet-semantics/README.md (hunk 1 above)
Comment (cwgoes) [reply spec/ics-004-channel-and-packet-semantics/README.md]: Not necessary; it depends on the details of how we store the routing table.

For now it is true though.

File: spec/ics-004-channel-and-packet-semantics/README.md [hunk 6]
```diff
@@ -67,18 +67,19 @@ interface ChannelEnd {
...
 
 - The `state` is the current state of the channel end.
+- The `ordering` field indicates whether the channel is ordered or unordered.
 - The `counterpartyChannelIdentifier` identifies the channel end on the counterparty chain.
-- The `portIdentifier` identifies the module which owns this channel end.
-- The `counterpartyPortIdentifier` identifies the module on the counterparty chain which owns the other end of the channel.
+- The `counterpartyPortIdentifier` identifies the port on the counterparty chain which owns the other end of the channel.
 - The `nextSequenceSend`, stored separately, tracks the sequence number for the next packet to be sent.
 - The `nextSequenceRecv`, stored separately, tracks the sequence number for the next packet to be received.
+- The `connectionHops` stores the list of connection identifiers, in order, along which packets sent on this channel will travel. At the moment this list must be of length 2.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply spec/ics-004-channel-and-packet-semantics/README.md]: Added a note; updated visualization.

**Pull request merged**